packages/block-library/src/heading/test/edit.native.js

```diff
@@ -0,0 +1,111 @@
+/**
+ * External dependencies
+ */
+import { shallow } from 'enzyme';
+
+/**
+ * WordPress dependencies
+ */
+import * as data from '@wordpress/data';
+import * as blockLibrary from '@wordpress/block-library';
+import * as blocks from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import HeadingToolbar from '../heading-toolbar';
+import HeadingEdit from '../edit';
+
+const blockEditorSelect = data.select( 'core/block-editor' );
+const blockEditorDispatch = data.dispatch( 'core/block-editor' );
+
+const testAttributes = {
+	content: 'Heading block',
+	level: 4,
+	align: 'center',
+	style: { color: { text: '#003fa3' } },
+};
+
+describe( 'Heading', () => {
+	beforeAll( () => {
+		blockLibrary.registerCoreBlocks();
+		gutenbergSetup();
+		blockEditorDispatch.removeBlocks( blockEditorSelect.getBlocks() );
+	} );
+
+	it( 'Serializes all Heading attributes properly', () => {
+		const expectedHTML = `<!-- wp:heading {"align":"center","level":4,"style":{"color":{"text":"#003fa3"}}} -->
+<h4 class="has-text-align-center has-text-color" style="--wp--color--text:#003fa3">Heading block</h4>
+<!-- /wp:heading -->`;
+
+		insertHeadingBlockWithAttributes( testAttributes );
+
+		const finalHTML = blocks.serialize( blockEditorSelect.getBlocks() );
+
+		expect( finalHTML ).toEqual( expect.stringContaining( expectedHTML ) );
+	} );
+
+	it( 'Parse Heading attributes from HTML properly', () => {
+		const expectedHTML = `<!-- wp:heading {"align":"center","level":4,"style":{"color":{"text":"#003fa3"}}} -->
+<h4 class="has-text-align-center has-text-color" style="--wp--color--text:#003fa3">Heading block</h4>
+<!-- /wp:heading -->`;
+
+		const parsedBlocks = blocks.parse( expectedHTML );
+
+		expect( parsedBlocks[ 0 ].attributes ).toEqual( testAttributes );
+	} );
+
+	it( 'Changes Heading level via HeaderToolBar ', () => {
+		const expectedHTML = `<!-- wp:heading {"level":4} -->
+<h4>Heading block</h4>
+<!-- /wp:heading -->`;
+
+		const attributes = {
+			content: 'Heading block',
+		};
+
+		const headingEdit = createHeadingEdit( attributes );
+		headingEdit
+			.find( HeadingToolbar )
```

Comment:
Interesting. This seems to try to drill down inside the block's component hierarchy to try and "tap" on the heading level-4 by calling the component's `onChange`. I guess that assumes quite a few assumptions about the code and hierarchy. This feels quite fragile, especially if this test is not shared with the web.

And I guess it brings up the question: Can the blocks be "unit-tested"? And if so, how? What's the public API of any particular block? Feels like we should only test using such a public API, if any, otherwise we end up with drilling too deep into the component hierarchy?

I wonder if the blocks are unit-tested on the web codepath or not. For the Heading block, I don't see any meaningful unit test in [its folder](https://github.com/WordPress/gutenberg/tree/ddd3fffdd327d2f9c5202481345fb7427d4a822b/packages/block-library/src/heading/test), though there are [more than a few e2e tests for it](https://github.com/WordPress/gutenberg/blob/6c95d81bba42681ee4e1f60ba4414bdde4167a3d/packages/e2e-tests/specs/editor/blocks/heading.test.js). The majority of the blocks on the web codepath don't seem to have unit tests.

Reply:
The idea is to trigger the native code paths from the nearest to the block as it is possible. Maybe the public block api would be the `attributes` and `setAttributes` props, as in the previous test examples, but those won't run code from the `edit(.native).js` files.

Ideally we will make the Edit component render to check if some import is missing on mobile, or some other external change affected it.

Maybe this test breaks because `HeadingToolbar` has been replaced by something else. And that something else might not be implemented yet on Mobile, so it's a good signal for us. Best case we just need to update the test.

For this reason I think this kind of tests are still good for us, and at the same time probably not needed for web.